bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -822,10 +831,10 @@ public void run()
       };
     }
 
-    protected void release(boolean wait)
+    protected void release(boolean wait, boolean writer)
     {
       final int refCount = this.refCount.decrementAndGet();
-      if (refCount == 0 && storage != null) {
+      if (refCount == 0 && storage != null && canRelease(writer)) {
```

Comment:
@sandeshh regarding your second point, what you are suggesting should probably happen at the input operator because we need to maintain window data consistency. It would need to happen outside the buffer server but using buffer server stats.